### PR TITLE
feat: treeshake for optional chaining

### PR DESCRIPTION
**Pull request by antfu**
This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

### Description

esbuild treeshakes it as well: https://esbuild.egoist.dev/#W1siaW5kZXgudHMiLHsiY29udGVudCI6ImlmICh1bmRlZmluZWQ/LmEpIHtcbiAgY29uc29sZS5sb2coXCJcIik7XG59XG5cbnVuZGVmaW5lZD8uYSgpO1xuIn1dLFsiZXNidWlsZC5jb25maWcuanNvbiIseyJjb250ZW50Ijoie1xuICBcImZvcm1hdFwiOiBcImVzbVwiLFxuICBcImNkblVybFwiOiBcImh0dHBzOi8vY2RuLnNreXBhY2suZGV2XCIsXG4gIFwibWluaWZ5XCI6IHRydWVcbn0ifV0sWyJzdW0udHMiLHsiY29udGVudCI6ImV4cG9ydCBjb25zdCBzdW0gPSAoYTogbnVtYmVyLCBiOiBudW1iZXIpID0+IGEgKyBiIn1dXQ==

### CODE REVIEW

**Review by lukastaegert — APPROVED** [review]
Nice, great work!

File: src/ast/nodes/ChainExpression.ts [hunk 1]
```diff
@@ -1,9 +1,50 @@
...
 	declare type: NodeType.tChainExpression;
+	private objectValue: LiteralValueOrUnknown | typeof unset = unset;
+
+	deoptimizeCache(): void {
+		this.objectValue = UnknownValue;
+	}
+
+	getLiteralValueAtPath(): LiteralValueOrUnknown {
+		if (this.getObjectValue() == null) return undefined;
+		return UnknownValue;
```
Comment (lukastaegert) [inline src/ast/nodes/ChainExpression.ts:27]: This one should be easy to cover in a test, I wonder why you current test does not hit this line

File: src/ast/nodes/ChainExpression.ts [hunk 2]
```diff
@@ -1,9 +1,50 @@
...
+	}
+
+	hasEffects(context: HasEffectsContext): boolean {
+		if (this.getObjectValue() == null) return false;
+		return this.expression.hasEffects(context);
+	}
+
+	render(code: MagicString, options: RenderOptions): void {
```
Comment (lukastaegert) [inline src/ast/nodes/ChainExpression.ts:35]: You do not need the special rendering logic because the actual removal of code happens in the Nodes that can actually have children removed, i.e. `Program`, `BlockStatement` and some edge cases like `IfStatement`. The important part is the `hasEffects`, which can mark this expression as free of side effects.

File: src/ast/nodes/ChainExpression.ts [hunk 3]
```diff
@@ -1,9 +1,50 @@
...
-export default class ChainExpression extends NodeBase {
+const unset = Symbol('unset');
+
+export default class ChainExpression extends NodeBase implements DeoptimizableEntity {
 	declare expression: CallExpression | MemberExpression;
 	declare type: NodeType.tChainExpression;
+	private objectValue: LiteralValueOrUnknown | typeof unset = unset;
+
+	deoptimizeCache(): void {
+		this.objectValue = UnknownValue;
```
Comment (lukastaegert) [inline src/ast/nodes/ChainExpression.ts:22]: This one is actually hard to cover in a test and I am fine if it is not covered. To hit this, you have to use a variable further up in the chain that is initialized as `undefined` but later gets another value assigned, invalidating the `object.getLiteralValueAtPath` in line 46.
The problem is to find a suitable "later". Sometimes it works to do this in a function but as I said, it is tricky.

File: src/ast/nodes/ExpressionStatement.ts [hunk 4]
```diff
@@ -1,14 +1,24 @@
...
 export default class ExpressionStatement extends StatementBase {
 	declare directive?: string;
 	declare expression: ExpressionNode;
 
+	getLiteralValueAtPath(
+		path: ObjectPath,
+		recursionTracker: PathTracker,
+		origin: DeoptimizableEntity
+	) {
+		return this.expression.getLiteralValueAtPath(path, recursionTracker, origin);
```
Comment (lukastaegert) [inline src/ast/nodes/ExpressionStatement.ts:19]: Why did you add this one? I cannot think of a scenario where the literal value of an ExpressionStatement would be queried as it cannot be nested in other expressions except `BlockStatement` and that one does not care about values. Maybe we can remove it.

Comment (antfu) [reply src/ast/nodes/ExpressionStatement.ts]: Removed. I wasn't familiar the pipeline so I tried different places :P

**Comment by antfu** [comment]
Updated. Thanks for the guidance! I will leave the coverage `deoptimizeCache` as I don't exactly understand the context and how. Feel free to push. Thanks!

**Review by lukastaegert — APPROVED** [review]
Great stuff, thanks!

**Comment by lukastaegert** [comment]
This PR has been released as part of rollup@3.10.0. You can test it via `npm install rollup`.
